### PR TITLE
Upgrade to Google Drive v3  [#171392351]

### PR DESCRIPTION
**Pull request by dougmartin**
Upgraded from v2 to V3 of the Google API.  This upgrade requires a new apiKey option in the config.  This change also added two new strings that needs to be internationalized.

The local secure test port was changed from 8888 to 8080 as that it what is defined in the Google Developer console.

### CODE REVIEW

File: src/code/providers/google-drive-provider.js [hunk 1]
```diff
@@ -101,19 +93,27 @@ class GoogleDriveProvider extends ProviderInterface {
...
       throw new Error((tr("~GOOGLE_DRIVE.ERROR_MISSING_CLIENTID")))
     }
-    this.scopes = this.options.scopes || [
+    this.scopes = (this.options.scopes || [
       'https://www.googleapis.com/auth/drive',
       'https://www.googleapis.com/auth/drive.install',
       'https://www.googleapis.com/auth/drive.file',
       'https://www.googleapis.com/auth/userinfo.profile'
-    ]
+    ]).join(" ")
```
Comment (dougmartin) [inline src/code/providers/google-drive-provider.js:109]: Yes, between v2 and v3 they changed the scopes from an array to a space delimited string (with no documentation about making this change!)

File: src/code/providers/google-drive-provider.js [hunk 2]
```diff
@@ -57,13 +40,22 @@ const GoogleDriveAuthorizationDialog = createReactClassFactory({
...
+      case "not-loaded":
+        contents = tr("~GOOGLE_DRIVE.CONNECTING_MESSAGE")
+        break
+      case "loaded":
+        contents = button({onClick: this.authenticate}, (tr("~GOOGLE_DRIVE.LOGIN_BUTTON_LABEL")))
+        break
+      case "errored":
+        contents = tr("~GOOGLE_DRIVE.ERROR_CONNECTING_MESSAGE")
+        break
+    }
```
Comment (kswenson) [inline src/code/providers/google-drive-provider.js:54]: Could be written more compactly as a map lookup:
```javascript
const contents = {
  "not-loaded": tr("~GOOGLE_DRIVE.CONNECTING_MESSAGE"),
  "loaded": button({onClick: this.authenticate}, (tr("~GOOGLE_DRIVE.LOGIN_BUTTON_LABEL"))),
  "errored": tr("~GOOGLE_DRIVE.ERROR_CONNECTING_MESSAGE")
}[this.state.gapiLoadState]
```

**Review by kswenson — APPROVED** [review]
👍 Looks good to me! I'm not in a position to validate the API changes themselves but the code changes make sense. I made a few stylistic suggestions but nothing substantive. Also, it might be useful to include comments with any documentation links you found useful or from which you took sample code for the benefit of those following in your footsteps.

**Comment by kswenson** [comment]
Not sure why it duplicated some of my comments, but you can ignore the extras.

File: src/code/providers/google-drive-provider.js [hunk 3]
```diff
@@ -277,78 +284,51 @@ class GoogleDriveProvider extends ProviderInterface {
...
-        // recent change in the behavior of the api is causing that error handler not to
-        // be called
-        const url = file.downloadUrl.replace(/^https:\/\/content\.googleapis\.com/, "https://www.googleapis.com")
-        return download(url)
-      } else {
-        return callback(this._apiError(file, 'Unable to get download url'))
+        })
       }
+      const xhr = new XMLHttpRequest()
+      xhr.open('GET', `https://www.googleapis.com/drive/v3/files/${metadata.providerData.id}?alt=media`)
```
[the remark this answers is not in the document]
Comment (dougmartin) [reply src/code/providers/google-drive-provider.js:327]: The actually removed the explicit downloadUrl in v3 file get results.  There is an export method for server side node to download files - this is the url that it generates.  Its a bit of a risk to use it but we need to do it.

**Comment by dougmartin** [comment]
@kswenson I added the download url comment as requested.  I'm going to leave this PR open so that it can be tested as a branch in SageModeler.  Once tested I'll merge it.

**Comment by scytacki** [comment]
@dougmartin or @kswenson what is the status of this. I thought that CODAP is already using this update, but it seems like it hasn't been merged. 

Is CODAP using code from this branch instead of master?

**Comment by dougmartin** [comment]
@scytacki I'm not sure of the CODAP status.  Now that we are back from the holidays I think it makes sense to merge it.

**Comment by kswenson** [comment]
>@dougmartin or @kswenson what is the status of this. I thought that CODAP is already using this update, but it seems like it hasn't been merged.

I believe CODAP is using `master` now, but @jsandoe has been managing CODAP/CFM releases on the CODAP side so he'd be the authority.

**Comment by jsandoe** [comment]
@kswenson, yes, CODAP is pulling from CFM master.

**Pull request merged**